File: algorithms/cpp/139.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
private:
    int binarySearch(vector<string> &wordDict, const string &word) {
        int l = 0, r = wordDict.size()-1;
        while (l <= r) {
            int m = l + (r-l)/2;
            if (wordDict[m] == word)
                return m;
            else if (wordDict[m] < word)
                l = m+1;
            else
                r = m-1;
        }
        return -1;
    }
public:
    bool wordBreak(string s, vector<string> &wordDict) {
        int n = s.size();

        sort(wordDict.begin(), wordDict.end());

        vector<bool> dp(n+1, false);
        dp[0] = true;

        for (int i = 1; i <= n; i++) {
            for (int j = i-1; j >= 0; j--) {
                if (dp[j] && binarySearch(wordDict, s.substr(j, i-j)) >= 0) {
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[n];
    }
};

int main() {
    vector<string> wordDict = {"a", "b"};
    string s = "ab";

    Solution solution;
    solution.wordBreak(s, wordDict);
    return 0;
}

File: algorithms/python/139.py
class TrieNode:
    def __init__(self):
        self.data = {}
        self.isWord = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for c in word:
            if c not in node.data.keys():
                node.data[c] = TrieNode()
            node = node.data[c]
        node.isWord = True

    def search(self, word):
        node = self.root
        for c in word:
            if c not in node.data.keys():
                return False
            node = node.data[c]
        return node.isWord

class Solution:
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        trie = Trie()
        for word in wordDict:
            trie.insert(word)

        dp = [False] * (len(s)+1)
        dp[0] = True
        for i in range(len(dp)):
            if not dp[i]:
                continue
            for j in range(i+1, len(dp)):
                if trie.search(s[i:j]):
                    dp[j] = True
        return dp[-1]
